File: app/src/main/java/com/ty/zbpet/ui/activity/material/ArrivalInTodoDetailActivity.java
package com.ty.zbpet.ui.activity.material;

import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.util.SparseArray;
import android.view.KeyEvent;
import android.view.View;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.TextView;

import com.bigkoo.pickerview.listener.OnTimeSelectListener;
import com.pda.scanner.ScanReader;
import com.pda.scanner.Scanner;
import com.ty.zbpet.R;
import com.ty.zbpet.bean.CarPositionNoData;
import com.ty.zbpet.bean.ResponseInfo;
import com.ty.zbpet.bean.material.MaterialDetailsIn;
import com.ty.zbpet.bean.material.MaterialTodoSave;
import com.ty.zbpet.net.HttpMethods;
import com.ty.zbpet.presenter.material.MaterialPresenter;
import com.ty.zbpet.presenter.material.MaterialUiObjInterface;
import com.ty.zbpet.ui.MainApp;
import com.ty.zbpet.ui.adapter.material.MaterialTodoDetailAdapter;
import com.ty.zbpet.ui.base.BaseActivity;
import com.ty.zbpet.ui.widght.SpaceItemDecoration;
import com.ty.zbpet.util.ACache;
import com.ty.zbpet.util.CodeConstant;
import com.ty.zbpet.util.DataUtils;
import com.ty.zbpet.util.ResourceUtil;
import com.ty.zbpet.util.ZBLog;
import com.ty.zbpet.util.ZBUiUtils;
import com.ty.zbpet.util.scan.ScanBoxInterface;
import com.ty.zbpet.util.scan.ScanObservable;
import com.zhouyou.http.exception.ApiException;
import com.zhouyou.http.subsciber.BaseSubscriber;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Locale;

import butterknife.BindView;
import io.reactivex.disposables.Disposable;
import okhttp3.RequestBody;

/**
 * 原辅料——到货入库详情、待办
 * <p>
 * MaterialUiObjInterface<MaterialTodoDetailsData> ：数据接口
 * MaterialTodoDetailAdapter.SaveEditListener      ：输入框接口
 * ScanBoxInterface                                ：扫码接口
 *
 * @author TY
 */
public class ArrivalInTodoDetailActivity extends BaseActivity implements MaterialUiObjInterface<MaterialDetailsIn>
        , MaterialTodoDetailAdapter.SaveEditListener, ScanBoxInterface {

    @BindView(R.id.rv_in_storage_detail)
    RecyclerView recyclerView;
    @BindView(R.id.tv_time)
    TextView tvTime;
    @BindView(R.id.et_desc)
    EditText etDesc;

    private MaterialTodoDetailAdapter adapter;
    private List<MaterialDetailsIn.ListBean> list = new ArrayList<>();

    private String sapOrderNo;
    private String supplierId; // 供应商 ID
    private String warehouseId;

    /**
     * 时间选择
     */
    private String selectTime;

    /**
     * 点击库位码输入框
     */
    private Boolean currentFocus = false;

    /**
     * list 中 Position
     */
    private int currentPosition = -1;

    private Disposable disposable;
    private Scanner scanner = ScanReader.getScannerInstance();
    private ScanObservable scan = new ScanObservable(this);

    /**
     * 保存用户在输入框中的数据
     */
    private SparseArray<String> bulkNumArray = new SparseArray(10);
    private SparseArray<String> carCodeArray = new SparseArray(10);
    private SparseArray<String> batchNoArray = new SparseArray(10);

    /**
     * 库位码 ID
     */
    private SparseArray<String> positionId = new SparseArray(10);

    private MaterialPresenter materialPresenter = new MaterialPresenter(this);

    @Override
    protected void onBaseCreate(Bundle savedInstanceState) {
        SimpleDateFormat sdf = new SimpleDateFormat(CodeConstant.DATE_SIMPLE_H_M, Locale.CHINA);
        selectTime = sdf.format(new Date());
        tvTime.setText(selectTime);

    }

    @Override
    protected int getActivityLayout() {
//        return R.layout.activity_arrival_in_storage_detail;
        return R.layout.activity_content_row_two;
    }

    @Override
    protected void initOneData() {

        sapOrderNo = getIntent().getStringExtra("sapOrderNo");
        supplierId = getIntent().getStringExtra("supplierId");

        materialPresenter.fetchTODOMaterialDetails(sapOrderNo);
    }

    @Override
    protected void initTwoView() {

        initToolBar(R.string.label_purchase_detail, new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                // 冲销入库
                doPurchaseInRecallOut(initParam());

            }
        });

        findViewById(R.id.add_ship).setVisibility(View.GONE);
        TextView titleName = findViewById(R.id.in_storage_detail);
        titleName.setText("到货明细");

        tvTime.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {

                ZBUiUtils.showPickDate(ArrivalInTodoDetailActivity.this, new OnTimeSelectListener() {
                    @Override
                    public void onTimeSelect(Date date, View v) {
                        //选中事件回调
                        selectTime = ZBUiUtils.getTime(date);
                        tvTime.setText(selectTime);
                        ZBUiUtils.showToast(selectTime);
                    }
                });
            }
        });
    }


    private RequestBody initParam() {

        MaterialTodoSave requestBody = new MaterialTodoSave();
        List<MaterialTodoSave.DetailsBean> detail = new ArrayList<>();

        int size = list.size();
        for (int i = 0; i < size; i++) {
            String bulkNum = bulkNumArray.get(i);
            String carCode = carCodeArray.get(i);
            String batchNo = batchNoArray.get(i);

            String Id = positionId.get(i);

            MaterialTodoSave.DetailsBean bean = new MaterialTodoSave.DetailsBean();
            if (null != bulkNum && null != carCode) {
                if (!bulkNum.isEmpty()) {
                    bean.setNumber(bulkNum);
                } else if (!carCode.isEmpty()) {
                    bean.setPositionId(carCode);
                } else if (!batchNo.isEmpty()) {
                    bean.setSapMaterialBatchNo(batchNo);
                }
                bean.setPositionId(Id);
                bean.setSupplierId(supplierId);
                bean.setSupplierNo(list.get(i).getSupplierNo());
                bean.setMaterialId(list.get(i).getMaterialId());
                bean.setConcentration(list.get(i).getConcentration());

                detail.add(bean);
            } else {
                // 不处理
                continue;
            }
        }

        // 没有合法的操作数据,不请求网络
        if (detail.size() == 0) {
            ZBUiUtils.showToast("请完善您要保存的信息");
            return null;
        }
        String remark = etDesc.getText().toString().trim();
        String time = tvTime.getText().toString().trim();

        requestBody.setDetails(detail);
        requestBody.setWarehouseId(warehouseId);
        requestBody.setInTime(time);
        requestBody.setSapOrderNo(sapOrderNo);
        requestBody.setRemark(remark);


        String json = DataUtils.toJson(requestBody, 1);
        ZBLog.e("JSON " + json);
        return RequestBody.create(okhttp3.MediaType.parse("application/json;charset=UTF-8"), json);
    }

    /**
     * 入库保存
     *
     * @param body
     */
    private void doPurchaseInRecallOut(RequestBody body) {

        HttpMethods.getInstance().materialTodoInSave(new BaseSubscriber<ResponseInfo>() {
            @Override
            public void onError(ApiException e) {
                ZBUiUtils.showToast(e.getMessage());
            }

            @Override
            public void onNext(ResponseInfo responseInfo) {
                if (CodeConstant.SERVICE_SUCCESS.equals(responseInfo.getTag())) {
                    // 入库成功（保存）
                    ZBUiUtils.showToast(responseInfo.getMessage());
                    runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            finish();
                        }
                    });
                } else {
                    ZBUiUtils.showToast(responseInfo.getMessage());
                }
            }
        }, body);
    }

    @Override
    public void detailObjData(MaterialDetailsIn obj) {

        list.clear();
        list.addAll(obj.getList());


        if (adapter == null) {
            LinearLayoutManager manager = new LinearLayoutManager(ResourceUtil.getContext());
            recyclerView.addItemDecoration(new SpaceItemDecoration(ResourceUtil.dip2px(10), false));
            recyclerView.setLayoutManager(manager);
            adapter = new MaterialTodoDetailAdapter(this, R.layout.item_matterial_todo_detail, obj.getList());
            recyclerView.setAdapter(adapter);

            adapter.setOnItemClickListener(new MaterialTodoDetailAdapter.OnItemClickListener() {
                @Override
                public void onItemClick(View view, RecyclerView.ViewHolder holder, int position) {

                    View rlDetail = holder.itemView.findViewById(R.id.rl_detail);
                    ImageView ivArrow = holder.itemView.findViewById(R.id.iv_arrow);

                    if (rlDetail.getVisibility() == View.VISIBLE) {
                        rlDetail.setVisibility(View.GONE);
                        ivArrow.setImageResource(R.mipmap.ic_collapse);
                    } else {
                        rlDetail.setVisibility(View.VISIBLE);
                        ivArrow.setImageResource(R.mipmap.ic_expand);

                    }

                    ZBUiUtils.hideInputWindow(ArrivalInTodoDetailActivity.this, view);

                }

                @Override
                public boolean onItemLongClick(View view, RecyclerView.ViewHolder holder, int position) {
                    return false;
                }
            });
        } else {
            adapter.notifyDataSetChanged();
        }
    }


    @Override
    public void saveEditAndGetHasFocusPosition(String etType, Boolean hasFocus, int position, EditText editText) {
        // 用户在 EditText 中输入的数据
        currentPosition = position;
        String textContent = editText.getText().toString().trim();

        if (CodeConstant.ET_ZKG.equals(etType)) {
            // 和 库位码 一样的存值方法
            MainApp.mCache.put(CodeConstant.ET_ZKG, position + "@" + textContent);

        } else if (CodeConstant.ET_BULK_NUM.equals(etType)) {
            bulkNumArray.put(position, textContent);

            // 库位码 需要判断合法性
        } else if (CodeConstant.ET_CODE.equals(etType)) {
            currentFocus = hasFocus;

            // 【情况 ② 】 无焦点 有内容 http 校验
            if (!TextUtils.isEmpty(textContent)) {
                if (!hasFocus) {
                    //httpCheckCarCode(currentPosition, textContent);
                }
            }

            carCodeArray.put(position, textContent);

        } else if (CodeConstant.ET_BATCH_NO.equals(etType)) {
            batchNoArray.put(position, textContent);

        }

    }


    /**
     * 扫描
     *
     * @param keyCode
     * @param event
     * @return
     */
    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {

        if (keyCode == CodeConstant.KEY_CODE_131
                || keyCode == CodeConstant.KEY_CODE_135
                || keyCode == CodeConstant.KEY_CODE_139) {

            if (currentFocus && currentPosition != -1) {
                // 扫描
                doDeCode();
            }


            return true;
        }
        return super.onKeyDown(keyCode, event);
    }


    private void doDeCode() {

        scanner.open(getApplicationContext());

        disposable = scan.scanBox(scanner, currentPosition);

    }

    /**
     * scan.scanBox 成功回调
     * 【情况 ① 】
     * 有焦点 扫码  http 校验
     *
     * @param positionNo
     */
    @Override
    public void ScanSuccess(int position, String positionNo) {
        ZBUiUtils.showToast("库位码 ：" + positionNo);
        //adapter.notifyItemChanged(position);

        //  服务器校验 库位码
        httpCheckCarCode(position, positionNo);

    }

    @Override
    public void showCarSuccess(int position, CarPositionNoData carData) {
        if (carData.getCount() > 0) {
            ZBUiUtils.showToast("扫码成功 === showCarSuccess ");
            String carId = carData.getList().get(0).getId();
            warehouseId = carData.getList().get(0).getWarehouseId();
            positionId.put(position, carId);

            adapter.notifyItemChanged(position);
        } else {
            ZBUiUtils.showToast("请扫正确的库位码");
        }
    }

    /**
     * Http 校验 库位码合法
     *
     * @param position   item 更新需要的 position
     * @param positionNo 扫码的编号
     */
    private void httpCheckCarCode(final int position, final String positionNo) {

        materialPresenter.checkCarCode(position, positionNo);

    }


    @Override
    protected void onDestroy() {
        super.onDestroy();

        scanner.close();

    }
}

File: app/src/main/java/com/ty/zbpet/ui/activity/material/ArrivalInTodoDetailActivity.kt
package com.ty.zbpet.ui.activity.material

import android.os.Bundle
import android.support.v7.util.DiffUtil
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.util.SparseArray
import android.view.KeyEvent
import android.view.View
import android.widget.*
import com.pda.scanner.ScanReader
import com.ty.zbpet.R
import com.ty.zbpet.bean.CarPositionNoData
import com.ty.zbpet.bean.material.MaterialDetails
import com.ty.zbpet.constant.CodeConstant
import com.ty.zbpet.data.DeepCopyData
import com.ty.zbpet.data.SharedP
import com.ty.zbpet.net.RequestBodyJson
import com.ty.zbpet.presenter.material.MaterialPresenter
import com.ty.zbpet.presenter.material.MaterialUiListInterface
import com.ty.zbpet.ui.adapter.diffadapter.TodoCarCodeDiffUtil
import com.ty.zbpet.ui.adapter.material.MaterialTodoDetailAdapter
import com.ty.zbpet.ui.base.BaseActivity
import com.ty.zbpet.ui.widght.ShowDialog
import com.ty.zbpet.ui.widght.SpaceItemDecoration
import com.ty.zbpet.util.DataUtils
import com.ty.zbpet.util.JsonStringMerge
import com.ty.zbpet.util.ResourceUtil
import com.ty.zbpet.util.ZBUiUtils
import com.ty.zbpet.util.scan.ScanBoxInterface
import com.ty.zbpet.util.scan.ScanObservable
import com.xiasuhuei321.loadingdialog.view.LoadingDialog
import com.zhy.adapter.recyclerview.MultiItemTypeAdapter
import io.reactivex.disposables.Disposable
import kotlinx.android.synthetic.main.activity_content_row_two.*
import okhttp3.RequestBody
import java.text.SimpleDateFormat
import java.util.*

/**
 * 原辅料——到货入库详情、待办
 * @author TY
 */
class ArrivalInTodoDetailActivity : BaseActivity()
        , MaterialUiListInterface<MaterialDetails.ListBean>
        , ScanBoxInterface {

    override val activityLayout: Int
        get() = R.layout.activity_content_row_two

    private lateinit var adapter: MaterialTodoDetailAdapter
    private var list = mutableListOf<MaterialDetails.ListBean>()

    private lateinit var sapOrderNo: String
    private lateinit var supplierNo: String
    private lateinit var supplierName: String
    private lateinit var creatorNo: String
    private lateinit var sapFirmNo: String
    private lateinit var content: String
    private lateinit var warehouseId: String
    private lateinit var warehouseNo: String
    /**
     * 时间选择
     */
    private lateinit var selectTime: String

    private lateinit var disposable: Disposable
    private val scanner = ScanReader.getScannerInstance()
    private val scan = ScanObservable(this)

    /**
     * 库位码 ID
     */
    private val positionId: SparseArray<String> = SparseArray(10)

    private val materialPresenter = MaterialPresenter(this)

    override fun onBaseCreate(savedInstanceState: Bundle?) {
        val sdf = SimpleDateFormat(CodeConstant.DATE_SIMPLE_H_M, Locale.CHINA)
        selectTime = sdf.format(Date())
        tv_time!!.text = selectTime

    }

    override fun initOneData() {

        sapFirmNo = intent.getStringExtra("sapFirmNo")
        sapOrderNo = intent.getStringExtra("sapOrderNo")
        supplierNo = intent.getStringExtra("supplierNo")
        supplierName = intent.getStringExtra("supplierName")
        creatorNo = intent.getStringExtra("creatorNo")
        content = intent.getStringExtra("content")

        materialPresenter.fetchTODOMaterialDetails(sapFirmNo, sapOrderNo, supplierNo)
    }

    override fun initTwoView() {

        initToolBar(R.string.label_purchase_detail, "保存", View.OnClickListener {
            ZBUiUtils.hideInputWindow(it.context, it)
            doPurchaseInRecallOut(initParam())
        })

        val titleName = findViewById<TextView>(R.id.in_storage_detail)
        titleName.text = "到货明细"

        tv_time!!.setOnClickListener {
            ZBUiUtils.showPickDate(it.context) { date, _ ->
                //选中事件回调
                selectTime = ZBUiUtils.getTime(date)
                tv_time!!.text = selectTime
                ZBUiUtils.showSuccess(selectTime)
            }
        }
    }


    private fun initParam(): RequestBody? {

        val requestBody = MaterialDetails()
        val detail = ArrayList<MaterialDetails.ListBean>()

        val size = list.size
        for (i in 0 until size) {

            val view = rv_in_storage_detail.getChildAt(i)

            var viewUnit = "KG"
            val radioGroup = view.findViewById<RadioGroup>(R.id.radioGroup)
            for (position in 0 until 3) {
                val radioButton = radioGroup.getChildAt(position) as RadioButton
                if (radioButton.isChecked) {
                    viewUnit = radioButton.text.toString()
                }
            }

            // 含量手输
            val concentration = view.findViewById<EditText>(R.id.tv_box_num).text.toString().trim { it <= ' ' }
            val viewCode = view.findViewById<EditText>(R.id.et_code).text.toString().trim { it <= ' ' }
            val viewNumber = view.findViewById<EditText>(R.id.et_bulk_num).text.toString().trim { it <= ' ' }
            val viewSap = view.findViewById<EditText>(R.id.et_batch_no).text.toString().trim { it <= ' ' }

            val id = positionId.get(i)

            val bean = MaterialDetails.ListBean()
            if (viewNumber.isNotEmpty() && viewCode.isNotEmpty()) {

                val subContent = list[i].content!!
                val mergeContent = JsonStringMerge().StringMerge(subContent, content)

                bean.concentration = concentration
                bean.number = viewNumber
                bean.positionNo = viewCode
                bean.sapMaterialBatchNo = viewSap
                bean.positionId = id
                bean.warehouseNo = list[i].warehouseNo
                bean.supplierNo = list[i].supplierNo
                bean.materialId = list[i].materialId

                bean.content = mergeContent
                bean.sapFirmNo = list[i].sapFirmNo
                bean.inNumber = list[i].inNumber
                bean.materialName = list[i].materialName
                bean.materialNo = list[i].materialNo
                bean.unit = viewUnit

                detail.add(bean)
            }
        }

        // 没有合法的操作数据,不请求网络
        if (detail.size == 0) {
            ZBUiUtils.showWarning("请完善您要入库的信息")
            return null
        }
        val remark = et_desc!!.text.toString().trim { it <= ' ' }
        val time = tv_time!!.text.toString().trim { it <= ' ' }

        requestBody.list = detail
//        requestBody.warehouseId = warehouseId
//        requestBody.warehouseNo = warehouseNo
        requestBody.supplierNo = supplierNo
        requestBody.supplierName = supplierName
        requestBody.creatorNo = creatorNo
        requestBody.moveType = "105"
        requestBody.inTime = time
        requestBody.sapOrderNo = sapOrderNo
        requestBody.remark = remark


        val json = DataUtils.toJson(requestBody, 1)
        return RequestBodyJson.requestBody(json)
    }

    /**
     * 入库保存
     *
     * @param body
     */
    private fun doPurchaseInRecallOut(body: RequestBody?) {

        if (body == null) {
            return
        }
        materialPresenter.materialTodoInSave(body)
    }

    override fun showMaterial(lists: MutableList<MaterialDetails.ListBean>?) {

        list = lists!!

        val manager = LinearLayoutManager(ResourceUtil.getContext())
        rv_in_storage_detail.addItemDecoration(SpaceItemDecoration(ResourceUtil.dip2px(CodeConstant.ITEM_DECORATION), false))
        rv_in_storage_detail.layoutManager = manager
        adapter = MaterialTodoDetailAdapter(this, R.layout.item_matterial_todo_detail, list)
        rv_in_storage_detail.adapter = adapter

        adapter.setOnItemClickListener(object : MultiItemTypeAdapter.OnItemClickListener {
            override fun onItemClick(view: View, holder: RecyclerView.ViewHolder, position: Int) {

                val rlDetail = holder.itemView.findViewById<View>(R.id.rl_detail)
                val ivArrow = holder.itemView.findViewById<ImageView>(R.id.iv_arrow)

                if (rlDetail.visibility == View.VISIBLE) {
                    rlDetail.visibility = View.GONE
                    ivArrow.setImageResource(R.mipmap.ic_collapse)
                } else {
                    rlDetail.visibility = View.VISIBLE
                    ivArrow.setImageResource(R.mipmap.ic_expand)
                }

                ZBUiUtils.hideInputWindow(view.context, view)
            }

            override fun onItemLongClick(view: View, holder: RecyclerView.ViewHolder, position: Int): Boolean {
                return false
            }
        })
    }

    /**
     * 扫描
     *
     * @param keyCode
     * @param event
     * @return
     */
    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {

        if (keyCode == CodeConstant.KEY_CODE_131
                || keyCode == CodeConstant.KEY_CODE_135
                || keyCode == CodeConstant.KEY_CODE_139) {
            if (SharedP.getFocus(this) && SharedP.getPosition(this) != -1) {
                // 扫描
                doDeCode()
            }
            return true
        }
        return super.onKeyDown(keyCode, event)
    }


    private fun doDeCode() {

        scanner?.open(applicationContext)

        disposable = scan.scanBox(scanner, SharedP.getPosition(this))

    }

    /**
     * scan.scanBox 成功回调
     * 【情况 ① 】
     * 有焦点 扫码  http 校验
     *
     * @param positionNo
     */
    override fun ScanSuccess(position: Int, positionNo: String) {

        //  服务器校验 库位码
        val warehouseNo = list[position].warehouseNo
        materialPresenter.checkCarCode(position, positionNo, warehouseNo)

    }

    override fun showCarSuccess(position: Int, carData: CarPositionNoData) {
        if (carData.count > 0) {
            val carId = carData.list!![0].id
            val positionNo = carData.list!![0].positionNo
            warehouseId = carData.list!![0].warehouseId!!
            // 扫库位码的 仓库编号
            warehouseNo = carData.list!![0].warehouseNo!!
            // 入库仓库编号
            val rawHouseNo = list[position].warehouseNo
            if (warehouseNo != rawHouseNo) {
                ZBUiUtils.showWarning("该库位是 $warehouseNo 请扫 $rawHouseNo 的库位码")
                return
            }
            positionId.put(position, carId)

            val deepCopyList = DeepCopyData.deepCopyList(list)

            deepCopyList[position].positionNo = positionNo

            val diffUtil = DiffUtil.calculateDiff(TodoCarCodeDiffUtil(list, deepCopyList))
            diffUtil.dispatchUpdatesTo(adapter)

        } else {
            ZBUiUtils.showWarning("请扫正确的库位码")
        }
    }

    override fun saveSuccess() {
        finish()
    }

    override fun showError(msg: String?) {
        ZBUiUtils.showError(msg)
    }

    private var dialog: LoadingDialog? = null
    override fun showLoading() {
        dialog = ShowDialog.showFullDialog(this@ArrivalInTodoDetailActivity, "保存中")
    }

    override fun hideLoading() {
        dialog?.close()
    }


    override fun onDestroy() {
        super.onDestroy()
        scanner?.close()
        SharedP.clearFocusAndPosition(this)

    }
}
